### PR TITLE
Make TableWriter report physicalWrittenBytes to OperatorStats

### PR DESCRIPTION
**Pull request by kewang1024**
For support of scaled table writer: https://github.com/facebookincubator/velox/issues/5665

In table writer operator, add function `updateWrittenBytes` to acquire the written 
data size from underlying dataSink.
This function is invoked whenever we add input or get output from table writer operator.

### CODE REVIEW

**Comment by kewang1024** [comment]
Reference code on java: https://github.com/prestodb/presto/blob/master/presto-spi/src/main/java/com/facebook/presto/spi/ConnectorPageSink.java

File: velox/connectors/hive/HiveDataSink.cpp [hunk 1]
```diff
@@ -395,6 +395,14 @@ void HiveDataSink::close() {
   }
 }
 
+int64_t HiveDataSink::getCompletedBytes() const {
+  int64_t completedBytes{0};
+  for (const auto& ioStats : ioStats_) {
+    completedBytes += ioStats->rawBytesWritten();
```
Comment (xiaoxmeng) [inline velox/connectors/hive/HiveDataSink.cpp:401]: Can you also leverage this to update "onDiskDataSizeInBytes" in fragment output with tests? Thanks!

File: velox/connectors/Connector.h [hunk 2]
```diff
@@ -142,6 +142,12 @@ class DataSink {
   virtual std::vector<std::string> finish() const = 0;
 
   virtual void close() = 0;
+
+  /// Gets the number of output bytes written by this data sink so far.
+  /// If size is not available, this method should return zero.
+  virtual int64_t getCompletedBytes() const {
```
Comment (xiaoxmeng) [inline velox/connectors/Connector.h:148]: Move getCompletedBytes above finish() and rename to writtenBytes();
```
  /// Returns the number of bytes written on disk by this data sink so far.
```

File: velox/connectors/hive/HiveDataSink.h [hunk 3]
```diff
@@ -388,6 +388,8 @@ class HiveDataSink : public DataSink {
 
   void close() override;
 
+  int64_t getCompletedBytes() const override;
```
Comment (xiaoxmeng) [inline velox/connectors/hive/HiveDataSink.h:391]: Could you add to test this in HiveDataSinkTest? Thanks!

File: velox/exec/TableWriter.cpp [hunk 4]
```diff
@@ -151,10 +152,16 @@ RowVectorPtr TableWriter::getOutput() {
   std::vector<VectorPtr> columns = {
       writtenRowsVector, fragmentsVector, commitContextVector};
 
+  updateWrittenBytes();
   return std::make_shared<RowVector>(
       pool(), outputType_, nullptr, numOutputRows, columns);
 }
 
+void TableWriter::updateWrittenBytes() {
+  auto lockedStats = stats_.wlock();
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.cpp:161]:
```
  const auto writtenBytes = dataSink_->getCompletedBytes();
  auto lockedStats = stats_.wlock();
  lockedStats->physicalWrittenBytes = writtenBytes;
```

File: velox/exec/TableWriter.h [hunk 5]
```diff
@@ -104,6 +104,8 @@ class TableWriter : public Operator {
  private:
   void createDataSink();
 
+  void updateWrittenBytes();
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.h:107]: Comment? And add to test the task stats in TableWriterTest? Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@kewang1024 thanks for the change % comments.

File: velox/connectors/Connector.h [hunk 6]
```diff
@@ -136,6 +136,12 @@ class DataSink {
   /// TODO maybe at some point we want to make it async.
   virtual void appendData(RowVectorPtr input) = 0;
 
+  /// Returns the number of bytes written on disk by this data sink so far.
+  /// If size is not available, this method should return zero.
```
Comment (xiaoxmeng) [inline velox/connectors/Connector.h:140]: Drop "/// If size is not available, this method should return zero."? We must support this? And put TODO in derived DataSink class implementation if not supported yet. Thanks!

File: velox/exec/TableWriter.h [hunk 7]
```diff
@@ -104,6 +104,9 @@ class TableWriter : public Operator {
  private:
   void createDataSink();
 
+  // Updates the OperatorStats physicalWrittenBytes with current written bytes
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.h:107]:
```
// Updates physicalWrittenBytes in the operator stats after adding each input.
```

File: velox/connectors/hive/HiveDataSink.cpp [hunk 8]
```diff
@@ -378,10 +386,10 @@ std::vector<std::string> HiveDataSink::finish() const {
               ("targetFileName", info->writerParameters.targetFileName())
               ("fileSize", 0)))
           ("rowCount", info->numWrittenRows)
-         // TODO(gaoge): track and send the fields when inMemoryDataSizeInBytes, onDiskDataSizeInBytes
+         // TODO(gaoge): track and send the fields when inMemoryDataSizeInBytes
          // and containsNumberedFileNames are needed at coordinator when file_renaming_enabled are turned on.
           ("inMemoryDataSizeInBytes", 0)
-          ("onDiskDataSizeInBytes", 0)
+          ("onDiskDataSizeInBytes", getWrittenBytes())
```
Comment (xiaoxmeng) [inline velox/connectors/hive/HiveDataSink.cpp:392]: Nice!

File: velox/exec/tests/TableWriteTest.cpp [hunk 9]
```diff
@@ -2007,6 +2007,7 @@ TEST_P(AllTableWriterTest, tableWriteOutputCheck) {
       numRows += obj["rowCount"].asInt();
       ASSERT_EQ(obj["updateMode"].asString(), "NEW");
 
+      ASSERT_GT(obj["onDiskDataSizeInBytes"].asInt(), 0);
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:2010]: Could you add to verify this value matches with file size on disk? Thanks!

File: velox/exec/TableWriter.h (hunk 7 above)
Comment (kewang1024) [reply velox/exec/TableWriter.h:107]: I will leave it as it is because this function can be called at multiple places, not only for adding input

**Review by xiaoxmeng — COMMENTED** [review]
@kewang1024 can you add PR description? Thanks!

File: velox/connectors/hive/HiveDataSink.cpp [hunk 10]
```diff
@@ -357,11 +357,20 @@ void HiveDataSink::computePartitionAndBucketIds(const RowVectorPtr& input) {
...
+  }
+  return completedBytes;
+}
+
 std::vector<std::string> HiveDataSink::finish() const {
   std::vector<std::string> partitionUpdates;
   partitionUpdates.reserve(writerInfo_.size());
 
-  for (const auto& info : writerInfo_) {
+  for (int i = 0; i < writerInfo_.size(); i++) {
```
Comment (xiaoxmeng) [inline velox/connectors/hive/HiveDataSink.cpp:372]:
```
nit: s/i++/++i/
```

File: velox/connectors/hive/HiveDataSink.cpp [hunk 11]
```diff
@@ -357,11 +357,20 @@ void HiveDataSink::computePartitionAndBucketIds(const RowVectorPtr& input) {
...
+    auto info = writerInfo_.at(i);
```
Comment (xiaoxmeng) [inline velox/connectors/hive/HiveDataSink.cpp:373]:
```
const auto& info = writerInfo_.at(i);
```

File: velox/connectors/hive/HiveDataSink.h [hunk 12]
```diff
@@ -454,6 +456,8 @@ class HiveDataSink : public DataSink {
   // writers_ are both indexed by partitionId.
   std::vector<std::shared_ptr<HiveWriterInfo>> writerInfo_;
   std::vector<std::unique_ptr<dwio::common::Writer>> writers_;
+  // IoStats for each writer
```
Comment (xiaoxmeng) [inline velox/connectors/hive/HiveDataSink.h:459]:
```
// IO statistics collected for each writer.
```

File: velox/exec/TableWriter.cpp [hunk 13]
```diff
@@ -151,10 +152,17 @@ RowVectorPtr TableWriter::getOutput() {
   std::vector<VectorPtr> columns = {
       writtenRowsVector, fragmentsVector, commitContextVector};
 
+  updateWrittenBytes();
```
Comment (xiaoxmeng) [inline velox/exec/TableWriter.cpp:155]: Move to l102 right after we set finished_ to true?

File: velox/exec/tests/TableWriteTest.cpp [hunk 14]
```diff
@@ -808,6 +805,11 @@ class TableWriteTest : public HiveConnectorTestBase {
     }
   }
 
+  int getNumWriters() {
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:808]: Can you update to all the call sites in this test file? Thanks!

File: velox/exec/tests/TableWriteTest.cpp [hunk 15]
```diff
@@ -2015,6 +2016,11 @@ TEST_P(AllTableWriterTest, tableWriteOutputCheck) {
       writeFiles.push_back(writeFileName);
       const std::string targetFileName =
           writerInfoObj["targetFileName"].asString();
+      const std::string writeFileFullPath =
+          obj["writePath"].asString() + "/" + writeFileName;
+      std::filesystem::path path{writeFileFullPath};
+      ASSERT_EQ(obj["onDiskDataSizeInBytes"].asInt(), file_size(path));
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:2022]: fs::file_size?
```
const auto actualFileSize = fs::file_size();
```

File: velox/exec/tests/TableWriteTest.cpp [hunk 16]
```diff
@@ -2051,6 +2057,53 @@ TEST_P(AllTableWriterTest, tableWriteOutputCheck) {
...
+      partitionKeys,
+      bucketProperty_,
+      numWriters,
+      connector::hive::LocationHandle::TableType::kNew,
+      commitStrategy_);
+
+  auto task = assertQueryWithWriterConfigs(
+      plan, inputFilePaths, "SELECT count(*) FROM tmp");
+  auto operatorStats = task->taskStats().pipelineStats.at(0).operatorStats;
+  for (int i = 0; i < operatorStats.size(); i++) {
```
Comment (xiaoxmeng) [inline velox/exec/tests/TableWriteTest.cpp:2100]: Can you add to check the result from TableWriteMerge operator as well? Thanks!

**Review by xiaoxmeng — COMMENTED** [review]
@kewang1024 thanks for the update % minors!

File: velox/exec/TableWriter.h (hunk 7 above)
Comment (xiaoxmeng) [reply velox/exec/TableWriter.h:107]: How about:
```
// Updates physicalWrittenBytes in OperatorStats with current written bytes.
```
Thanks!

**Review by xiaoxmeng — APPROVED** [review]
@kewang1024 thanks for the iterating!

**Pull request closed without merging**